Create Predeploy contracts (#3022)
OffchainLabs/arbitrum-docs: docs/launch-arbitrum-chain/02-configure-your-chain/common/use-a-custom-genesis.mdx

4. Reducing lift on 3rd-party infrastructure teams to redeploy contracts for new chains

## **What Does This Feature Do?**

This feature allows you to initialize a new blockchain with a customized starting state and network configuration through the following options:

- **Predeployed contracts** - standard chains start without any smart contracts. This feature allows you to preload contract bytecode in the very first block so infrastructure is available at the launch.
- **Initial account state (allocations)** - you can pre-configure the ledger—including account balances and contract storage—before the network opens for transactions.
- **Enable advanced features** – chain operators can use this feature to launch a chain with advanced customizations like [mint / burn of gas token via third party bridges](/launch-arbitrum-chain/features/common/gas-and-fees/choose-native-mint-burn) and compliance-focused transaction filtering.

## **How to use this feature?**

You can use the feature via the SDK or by using the `genesis-file-generator` tool.

### Use via chain SDK

:::caution

> The feature is supported in chain SDK versions xx.xx and above. Chain SDK only allows the standard `genesis.json` creation which contains the predeploy contracts. In case you want to customize the `genesis.json` file or pass accounts / custom chain config, you should use the `genesis-generator-tool` for using the feature.

Reviewer: ```suggestion
The feature is supported in chain SDK versions xx.xx and above. Chain SDK only allows the standard `genesis.json` creation which contains the predeploy contracts. In case you want to customize the `genesis.json` file or pass accounts / custom chain config, you should use the `genesis-file-generator` for using the feature.
```
genesis-generator-tool is a tool within nitro repo (Which is used to calculate block hash), however, the one we use to generate genesis file is genesis-file-generator

---
title: 'Use a custom genesis to deploy your Arbitrum chain'
sidebar_label: 'Use custom genesis'
description: 'Start your chain with a custom genesis file'
author: Mehta
sme: Mehta
user_story: As a current Arbitrum chain owner, I need to start my chain with a set of predeployed contracts and account balances using genesis.json file.
content_type: get-started
---

:::caution Minimum Requirement

- You must use Nitro contracts v3.2 or higher. This version is the first to include full support for custom `genesis.json` initialization.

> - Nitro node v3.10.0 or higher is needed for using custom-genesis.

Reviewer: ```suggestion
- Nitro node v3.9.9 or higher is needed for using custom-genesis.
```

| `ENABLE_NATIVE_TOKEN_SUPPLY` | _(Optional)_ Set to `true` if you want to launch your chain with native interop tokens as [gas token with mint / burn enabled](/launch-arbitrum-chain/features/common/gas-and-fees/choose-native-mint-burn) via third-party protocols. | `false` |
| `LOAD_DEFAULT_PREDEPLOYS` | _(Optional)_ Set to `false` if you don't want the default predeploys. | `true` |
| `ENABLE_TRANSACTION_FILTERING` | _(Optional)_ Set to `true` if you want to launch your chain with protocol-level transaction filtering for regulatory or compliance purposes. | `false` |

#### Execution Process

- **Prepare the genesis state:**

  - Set up the `.env` with the required parameters
  - Run the [offchainlabs/genesis-file-generator](https://hub.docker.com/r/offchainlabs/genesis-file-generator/tags) docker image to generate `genesis.json` file with the required pre-deployed contracts and additional configs.

    ```bash
    mkdir -p genesis

    docker run --rm \
    --env-file .env \
    -v "$(pwd)/genesis":/app/genesis \

>     offchainlabs/genesis-file-generator:v0.0.1-afad8d8

Reviewer: ```suggestion
    offchainlabs/genesis-file-generator:v0.0.2-4c37f62
```

| `ENABLE_NATIVE_TOKEN_SUPPLY` | _(Optional)_ Set to `true` if you want to launch your chain with native interop tokens as [gas token with mint / burn enabled](/launch-arbitrum-chain/features/common/gas-and-fees/choose-native-mint-burn) via third-party protocols. | `false` |
| `LOAD_DEFAULT_PREDEPLOYS` | _(Optional)_ Set to `false` if you don't want the default predeploys. | `true` |
| `ENABLE_TRANSACTION_FILTERING` | _(Optional)_ Set to `true` if you want to launch your chain with protocol-level transaction filtering for regulatory or compliance purposes. | `false` |

#### Execution Process

- **Prepare the genesis state:**

  - Set up the `.env` with the required parameters
  - Run the [offchainlabs/genesis-file-generator](https://hub.docker.com/r/offchainlabs/genesis-file-generator/tags) docker image to generate `genesis.json` file with the required pre-deployed contracts and additional configs.

    ```bash
    mkdir -p genesis

    docker run --rm \
    --env-file .env \
    -v "$(pwd)/genesis":/app/genesis \

>     offchainlabs/genesis-file-generator:latest

Reviewer: ```suggestion
    offchainlabs/genesis-file-generator:v0.0.1-afad8d8
```
We don't have 'latest' tag for this image

    ```

  - You can also generate your own genesis.json file, but carefully read [this notice](https://github.com/OffchainLabs/genesis-file-generator#exclamation-important-note-about-the-chain-config-property) before proceeding with the next steps.

- **Generate required hashes:** After the genesis.json file is created, run the nitro container to compute the genesis block hash with `genesis-generator` entrypoint:

  ```bash
  source .env

  docker run --rm \
    -v "$(pwd)/genesis":/data/genesisDir \
    --entrypoint genesis-generator \
    "$NITRO_NODE_IMAGE" \
    --genesis-json-file /data/genesisDir/genesis.json \

>     --initial-l1-base-fee "$L1_BASE_FEE"

Reviewer: ```suggestion
```
This flag is deleted
